Travis run the test in minimal and full config

jbarnoud opened the pull request:
This commit asks Travis to run the tests twice: once with all the requirements installed (SETUP=='full'), and once with none of the optional dependencies installed (SETUP=='minimal'). The minimal setup, has to run without the MKL as the conda mkl package pulls scipy with it.

The build fails which is expected. Indeed, until #568 get merged, some tests fail if scipy or netcdf are not installed.

Fix #483

kain88-de commented on .travis.yml:
```diff
@@ -43,10 +45,13 @@ after_success:
+    - GIT_CI_USER=TravisCI
+    - GIT_CI_EMAIL=TravisCI@mdanalysis.org
+    - MDA_DOCDIR=package/doc/html
+  matrix:
```
I prefer if this was at the top of the file where we define the python versions we use.

jbarnoud replied:
On 08/12/15 00:00, kain88-de wrote:
>
> In .travis.yml 
> <https://github.com/MDAnalysis/mdanalysis/pull/570#discussion_r46891970>:
>
> > \ No newline at end of file
> > +  global:
> > +    - secure="HIj3p+p2PV8DBVg/KGUx6n83KwB0ASE5FwOn0SMB9zxnzAqe8sapwdBQdMdq0s  XB7xT1spJqRxuxOMVEVn35BNLu7bxMLfa4287C8YXcomnvmv9xruxAsjsIewnNQ80vtPVbQddBPxa4  jKbqgPby5QhhAP8KANAqYe44pIV70fY="
> > +    - GH_DOC_BRANCH=develop
> > +    - GH_REPOSITORY=github.com/MDAnalysis/mdanalysis.git
> > +    - GIT_CI_USER=TravisCI
> > +    - GIT_CI_EMAIL=TravisCI@mdanalysis.org
> > +    - MDA_DOCDIR=package/doc/html
> > +  matrix:
>
> I prefer if this was at the top of the file where we define the python 
> versions we use.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/MDAnalysis/mdanalysis/pull/570/files#r46891970>.
>
I prefer it at the top too. But it was at the end so I left it there. I 
can move it up.

orbeckst replied:
By all means, move it up. I might have added it and I just followed other examples without much thinking.

richardjgowers commented:
This is awesome, thanks.  It looks like there's just a few things that need decorating with knownfail

kain88-de commented on .travis.yml:
```diff
@@ -17,16 +17,18 @@ before_install:
   - source activate pyenv
-  - conda install --yes python=$TRAVIS_PYTHON_VERSION cython biopython matplotlib networkx netcdf4
+  - if [[ $SETUP == 'full' ]]; then conda install --yes python=$TRAVIS_PYTHON_VERSION cython biopython matplotlib networkx netcdf4; fi
+  - if [[ $SETUP == 'minimal' ]]; then conda install --yes python=$TRAVIS_PYTHON_VERSION cython biopython networkx; fi
```
minimal should work without cython.

kain88-de replied:
Hm depending on when #560 get's merged maybe we can also say that a 'full' setup means we `cythonize` the files again and in the minimal setup we use the shipped `*.c` files.

jbarnoud commented:
I rebased against develop to account for #560. I also adapted the matrix to follow @richardjgowers suggestions in #560; 3 versions are built:

* full install with cythonize
* full install without cythonize
* minimal install without cythonize

I also removed cython from the packages installed in the minimal setup, and I moved the environment variable at the top of the file to improve readability.

kain88-de commented:
I don't think we need to run 3 versions. All the cython-compiled C-files will also be tested in the minimal install since none of them is optional (and they shouldn't be since then people who have installed cython but not the optional dependencies might have problems).

And also consider that we will have double the number of build-jobs as soon as we support python 3 as well.  With this in mind and travis asking nicely that we keep the build-matrix minimal I think we can skip **full install without cythonize**.

> Please take into account that Travis CI is an open source service and we rely on worker boxes provided by the community. So please only specify as big a matrix as you actually need.

jbarnoud commented:
I don't mind removing the full install without cythonize. It is just a matter of removing one line. What are the cases where the build with cythonize can pass while the build whithout fails? Or vice-versa?

kain88-de commented:
It can happen that someone doesn't update the cython generated C-files. If a bug was fixed in the pyx files that can lead to the tests failing for that bug.

richardjgowers commented:
I don't mind either way, so I'll let @kain88-de decide

jbarnoud commented:
If @kain88-de decides, then I remove the useless build :stuck_out_tongue:
